### PR TITLE
adds port to sqlcmd calls, plus tests

### PR DESCRIPTION
**Pull request by dzsquared**
While the entire connection string is passed directly to SqlPackage, because sqlcmd requires individual parameters (glare) we were dropping the port number from script execution and more importantly, connectivity testing.

This most prominent issue we saw was with SQL MI public endpoint connectivity, which uses 3342.

If no port is specified in the connection string, the default (1433) is used.

fyi @miljann995

I had an additional test of sqlcmd variables being picked up in sqlpackage calls in my local, so that's along for the ride.

secret value being used for public endpoint:
image

error on sql-action@v2.2 with debug logging enabled, shows it used 1433
image

same workflow with sql-action@drskwier/fix-server-port, connects successfully
image

### CODE REVIEW

**Review by zijchen — DISMISSED** [review]

File: src/SqlUtils.ts [hunk 1]
```diff
@@ -80,8 +80,8 @@ export default class SqlUtils {
             };
         }
         catch (error) {
-            core.debug(`${error.message}`);
-            core.debug(`SqlCmd stderr: ${sqlCmdError}`);
+            console.log(`${error.message}`);
```
Comment (chlafreniere) [inline src/SqlUtils.ts:83]: Where does core.debug output to? Does changing this surface areas more easily?

File: __tests__/SqlConnectionConfig.test.ts [hunk 2]
```diff
@@ -114,16 +114,18 @@ describe('SqlConnectionConfig tests', () => {
...
+            [`Server=test1.database.windows.net;Database=testdb;User Id=user;Password=placeholder;Authentication=SQLPassword`, 'test1.database.windows.net', '', 'Validates server name without prefix'],
+            [`Server=tcp:test1.database.windows.net;Database=testdb;User Id=user;Password=placeholder;Authentication=SQLPassword`, 'test1.database.windows.net', '', 'Validates server name with tcp prefix'],
+            [`Server=tcp:test1.database.windows.net,1433;Database=testdb;User Id=user;Password=placeholder;Authentication=SQLPassword`, 'test1.database.windows.net', '1433', 'Validates server name with tcp prefix and port'],
+            [`Server=test1.database.windows.net,1433;Database=testdb;User Id=user;Password=placeholder;Authentication=SQLPassword`, 'test1.database.windows.net', '1433', 'Validates server name with no prefix and port'],
+            [`Server=test1.database.windows.net,3342;Database=testdb;User Id=user;Password=placeholder;Authentication=SQLPassword`, 'test1.database.windows.net', '3342', 'Validates server name with no prefix and port'],
```
Comment (chlafreniere) [inline __tests__/SqlConnectionConfig.test.ts:121]: nit: perhaps we can make this and the above strings unique

File: src/SqlUtils.ts (hunk 1 above)
Comment (dzsquared) [reply src/SqlUtils.ts:83]: core.debug is only output if the user has selected the workflow to run with debug logging (https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging)

For most connectivity issue reports we're having to ask the users to go back and re-run the workflow with debug logging to get the basic connection failure information.  Changing this to console.log will output on any run that catches an error here.

File: src/SqlUtils.ts [hunk 3]
```diff
@@ -123,7 +123,7 @@ export default class SqlUtils {
                 throw new Error(`Platform ${process.platform} is not supported.`);
         }
 
-        let sqlcmdCall = `"${sqlCmdPath}" -S ${connectionConfig.server} -d ${connectionConfig.database}`;
+        let sqlcmdCall = `"${sqlCmdPath}" -S ${connectionConfig.server},${connectionConfig.port ?? 1433} -d ${connectionConfig.database}`;
```
[the remark this answers is not in the document]
Comment (dzsquared) [reply src/SqlUtils.ts:126]: I don't believe so - @shueybubbles is there anything with go-sqlcmd I might be missing about handling the port used when executing a script from this syntax?
I'm not aware of an environment variable or any other setting to change the default port it uses to connect.

[the remark this answers is not in the document]
Comment (shueybubbles) [reply src/SqlUtils.ts:126]: generally you would pass the `Server` string around "as is" and rely on all its consumers knowing to look for the port number in it. 

afaik passing 1433 is no other port was provided should be ok.

[the remark this answers is not in the document]
Comment (dzsquared) [reply src/SqlUtils.ts:126]: Unfortunately, this is one of those cases where users provide a connection string and while we can pass it directly to SqlPackage, we have to parse it up for sqlcmd.  We're using a node connection string parser to get the connection config parts, that usage is why server and port are handled independently.

**Review by miljann995 — APPROVED** [review]
Changes regarding the explicit port number in the `sqlcmd` command look good to me.

Regarding the SQL Managed Instance connectivity, I tried to visualize three main scenarios that we have here for the public endpoint with port `3342`:
diag

1. Regarding scenario A, that's a happy path, and it works:
image

2. Regarding the scenario B, that's the case where credentials are invalid, and this is the error message we get:
image

3. Regarding scenario C, that's the case where there's a lack of connectivity between SQL Managed Instance and GitHub server, and this is the error message we get:
image (2)

So, all in all, it works fine with these changes!

Just a side question - do we actually need to try adding a firewall rule if we try to connect to SQL Managed Instance? I guess that we can somehow figure out when to skip it since it will always fail in cases of SQL Managed Instance, so it's a redundant step in the end. The main motivation for this question is scenario C, where we don't have connectivity, but the error message suggests that there's no SQL Managed Instance of that name in the subscription. However, even if we choose to stick with this state of things, I guess we're fine.

Nevertheless, a big thanks for your changes and contribution!

**Review by zijchen — APPROVED** [review]

**Review by chlafreniere — APPROVED** [review]

**Comment by dzsquared** [comment]
merging in, but will consider how to smooth over the firewall rule component for SQLMI in the future

**Pull request merged**